File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        ArrayList<Integer> indexZero = new ArrayList<>();
        int[] ans=new int[nums.length];

        int product=1;
        for (int i=0; i<nums.length; i++) {
            if (nums[i]==0)
                indexZero.add(i);
            else
                product*=nums[i];
        }

        if (indexZero.size()==0) {
            for (int i=0; i<nums.length; i++) {
                ans[i]=product/nums[i];
            }
        }
        else if (indexZero.size()>1){
            for (int i=0; i<nums.length; i++)
                ans[i]=0;
        }
        else {
            for (int i=0; i<nums.length; i++) {
                if (i==indexZero.get(0)) {
                    ans[i]=product;
                }
                else
                    ans[i]=0;
            }
        }
        return ans;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ans = [0]*len(nums)
        index = -1
        zeros = 0
        product = 1

        for n in range(len(nums)):
            if nums[n]==0:
                zeros += 1
                if zeros==2:
                    return ans
                index = n
            else:
                product *= nums[n]
        
        if zeros==1:
            ans[index] = product
        else:
            for i in range(len(nums)):
                ans[i] = int(product / nums[i])
        return ans